File: God Zuo/Python/class060 拓扑排序的扩展技巧/My03.py
# 测试链接 : https://leetcode.cn/problems/parallel-courses-iii/
from collections import deque
from typing import List

class Solution:
    def minimumTime(self, n: int, relations: List[List[int]], time: List[int]) -> int:
        graph=[[] for i in range(n+1)]
        indegree=[0]*(n+1)
        for u,v in relations:
            graph[u].append(v)
            indegree[v]+=1
        queue=deque([ i for i in range(1,n+1) if indegree[i]==0 ])
        cost=[0]*(n+1)
        ans=0
        while queue:
            cur=queue.popleft()
            cost[cur]+=time[cur-1]
            ans=max(ans,cost[cur])
            for nxt in graph[cur]:
                cost[nxt]=max(cost[nxt],cost[cur])
                indegree[nxt]-=1
                if indegree[nxt]==0:
                    queue.append(nxt)
        return ans

File: God Zuo/class068 更多二维动态规划题目/My03.cpp
// 测试链接 : https://leetcode.cn/problems/interleaving-string/
#include<bits/stdc++.h>
using namespace std;

class Solution
{
public:
    bool isInterleave(string s1, string s2, string s3)
    {
        if(s1.length()+s2.length()!=s3.length())
            return false;
        bool dp[110][110];
        int n = s1.length();
        int m = s2.length();
        for (int i = 0; i <= n;++i){
            for (int j = 0; j <= m;++j){
                dp[i][j] = false;
            }
        }
            dp[0][0] = true;
        for (int j = 1; j <= m; ++j){
            if (s2[j-1]!=s3[j-1])
                break;
            dp[0][j] = true;
        }
        for (int i = 1; i <= n; ++i){
            if(s1[i - 1] != s3[i - 1])
                break;
            dp[i][0] = true;
        }
        for (int i = 1; i <= n;++i){
            for (int j = 1; j <= m;++j){
                dp[i][j] = (s1[i - 1] == s3[i + j - 1] && dp[i - 1][j]) || (s2[j - 1] == s3[i + j - 1] && dp[i][j - 1]);
            }
        }
        return dp[n][m];
    }
};